### PR TITLE
Bump zbus to 2.0.0-beta.7

### LINKED ISSUE #36: ZBus 2.0
Currently ZBus 2.0 is still in beta, but it'd probably not be a bad idea to start porting now so we can test the beta release before the final release.

### PR DESCRIPTION
**Pull request by Cogitri**
Fixes #36

### CODE REVIEW

**Comment by complexspaces** [comment]
Thanks for the PR to see what this would look like. Is there any ETA for when Zbus plans to do a make a stable release of 2.0? I was not able to find any tracking issue on the GitLab repository.

**Comment by Cogitri** [comment]
Not sure if there’s an ETA but I think it’ll get released soon since there have been 5 beta releases already

**Comment by zeenix** [comment]
Hi,

zbus maintainer here.

> Thanks for the PR to see what this would look like. Is there any ETA for when Zbus plans to do a make a stable release of 2.0?

Given that nobody works on zbus as their day job, it's hard to give a good estimate about 2.0 delivery day. What I can tell you is that we're very close to the API freeze. There may be API breakages before 2.0 but apart from dropping the proxy signal/property callback API (which this PR doesn't seem to be using directly at least), there are unlikely to be any major changes after beta.7.

**Comment by Cogitri** [comment]
Ah sorry, somehow I completely forgot to reply, @zeenix . Thanks for the estimate. I think at this point it"d be best to merge this into master and make a new release once zbus 2 is released.

**Comment by zeenix** [comment]
> Ah sorry, somehow I completely forgot to reply, @zeenix

NP. Actually, I'm hoping to release a API-freeze release today or tomorrow as there are no other API breakages planned. Then comes 2.0 in some days/weeks after people get a chance to test the API a bit more. There might still be API breakages before 2.0 but the chances are extremely low and will only happen if absolutely necessary.

**Comment by Cogitri** [comment]
Thanks, updated this PR to beta8

File: Cargo.toml
Comment (complexspaces) [inline Cargo.toml:16]: Please leave these crypto updates to another PR.

File: Cargo.toml
Comment (complexspaces) [inline Cargo.toml:21]: Ditto.

File: src/collection.rs
Comment (complexspaces) [inline src/collection.rs:39]: Can this be [`from_static_str`](https://docs.rs/zbus_names/2.0.0/zbus_names/enum.BusName.html#method.from_static_str) instead?

File: src/collection.rs
Comment (complexspaces) [inline src/collection.rs:40]: Does this need to be owned? `.path()` appears to take a borrowed `ObjectPath`.

File: src/item.rs
Comment (complexspaces) [inline src/item.rs:35]: My previous comments also apply here.

File: src/util.rs
Comment (complexspaces) [inline src/util.rs:97]: Once more with that comment set 😅

File: src/util.rs
Comment (complexspaces) [inline src/util.rs:101]: Can this panic? If it can, we should return an error instead.

File: src/util.rs
Comment (complexspaces) [inline src/util.rs:115]: This TODO should remain here.

**Comment by complexspaces** [comment]
It looks like [version 2.0.0 has been released](https://crates.io/crates/zbus/2.0.0)! @zeenix thanks for the work that went into the release 🎉

@Cogitri Would you mind please updating this branch to use the stable version? I'll run CI again once you do. Please rebase onto `master` as well, which should fix the CI failures.

File: src/collection.rs
Comment (Cogitri) [reply src/collection.rs:40]: It *can* take a borrowed `ObjectPath`, but then we need to add the lifetimes to `Collection` too, which would require a bigger rebuild of other parts as well (since the `collection_path` currently is a temporary in `Session`

**Comment by Cogitri** [comment]
Seems like tests currently fail though, I'll look into that in a bit.

**Comment by zeenix** [comment]
> It looks like [version 2.0.0 has been released](https://crates.io/crates/zbus/2.0.0)! @zeenix thanks for the work that went into the release tada

You're most welcome! It was indeed a lot of hard work to get the async API sorted. I hope you're happy with the results. :)

File: src/util.rs
Comment (zeenix) [inline src/util.rs:108]: This method returns a Result so maybe `?` would be better here (might need a `.map_error(..)` though)? :thinking:

File: src/collection.rs
Comment (complexspaces) [reply src/collection.rs]: Ah, good point. This is fine then.

**Comment by Cogitri** [comment]
3 Tests still fail for me - but maybe that's just due to my system? AFAICS `secret-service-rs` doesn't mock the libsecret DBus-API so I think this might be my libsecret implementation acting up (?). Maybe someone else can try running the tests too. The output for me is:

```
failures:

---- item::test::should_create_and_delete_item stdout ----
thread 'main' panicked at 'item still existed', src/item.rs:174:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- item::test::should_get_and_set_item_attributes stdout ----
Attributes: {}
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `{}`,
 right: `{"test_attributes_in_item_get": "test"}`', src/item.rs:299:9

---- item::test::should_get_and_set_item_label stdout ----
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `"Test"`,
 right: `"Tester"`', src/item.rs:243:9


failures:
    item::test::should_create_and_delete_item
    item::test::should_get_and_set_item_attributes
    item::test::should_get_and_set_item_label

test result: FAILED. 23 passed; 3 failed; 5 ignored; 0 measured; 0 filtered out; finished in 1.42s

error: test failed, to rerun pass '--lib'
```

**Comment by complexspaces** [comment]
I took the test suite through a run in a fresh Ubuntu 20.04 instance and got similarly interesting results. With this branch, I got a total of 4 test failures:
```
---- item::test::should_create_and_delete_item stdout ----
thread 'item::test::should_create_and_delete_item' panicked at 'item still existed', src/item.rs:174:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- collection::test::should_search_items stdout ----
thread 'collection::test::should_search_items' panicked at 'called `Result::unwrap()` on an `Err` value: Zbus(MethodError(OwnedInterfaceName(InterfaceName(Str(Owned("org.freedesktop.DBus.Error.UnknownMethod")))), Some("No such interface “org.freedesktop.DBus.Properties” on object at path /org/freedesktop/secrets/collection/login/4"), Msg { type: Error, sender: UniqueName(Str(Borrowed(":1.45"))), reply-serial: 20, body: Signature: [
        s (115),
] }))', src/collection.rs:274:49

---- item::test::should_get_and_set_item_label stdout ----
thread 'item::test::should_get_and_set_item_label' panicked at 'assertion failed: `(left == right)`
  left: `"Test"`,
 right: `"Tester"`', src/item.rs:243:9

---- item::test::should_get_and_set_item_attributes stdout ----
Attributes: {}
thread 'item::test::should_get_and_set_item_attributes' panicked at 'assertion failed: `(left == right)`
  left: `{}`,
 right: `{"test_attributes_in_item_get": "test"}`', src/item.rs:299:9


failures:
    collection::test::should_search_items
    item::test::should_create_and_delete_item
    item::test::should_get_and_set_item_attributes
    item::test::should_get_and_set_item_label
```

On `master`, I get this failure instead:
```
---- item::test::should_get_and_set_item_attributes stdout ----
Attributes: {"xdg:schema": "org.freedesktop.Secret.Generic", "test_attributes_in_item_get": "test"}
thread 'item::test::should_get_and_set_item_attributes' panicked at 'assertion failed: `(left == right)`
  left: `{"xdg:schema": "org.freedesktop.Secret.Generic", "test_attributes_in_item_get": "test"}`,
 right: `{"test_attributes_in_item_get": "test"}`', src/item.rs:283:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    item::test::should_get_and_set_item_attributes
```

So, in total, there's a net gain of 3 failures. I think that some of these, especially the unknown method error, should be looked at before this is merged as it appears to be a regression (even in the face of these tests being shaky).

**Comment by Cogitri** [comment]
```
---- collection::test::should_search_items stdout ----
thread 'collection::test::should_search_items' panicked at 'called `Result::unwrap()` on an `Err` value: Zbus(MethodError(OwnedInterfaceName(InterfaceName(Str(Owned("org.freedesktop.DBus.Error.UnknownMethod")))), Some("No such interface “org.freedesktop.DBus.Properties” on object at path /org/freedesktop/secrets/collection/login/4"), Msg { type: Error, sender: UniqueName(Str(Borrowed(":1.45"))), reply-serial: 20, body: Signature: [
        s (115),
] }))', src/collection.rs:274:49
```

That one is especially weird - are you sure that the keyring is actual functional on that Ubuntu install? Seems like others only get the same DBus error when SSH'ing into the machine, see e.g. https://github.com/jaraco/keyring/issues/514#issuecomment-932863658

**Comment by Cogitri** [comment]
Oh also - are you running with multiple test thread? That seems to result in similar errors at random for me.

**Comment by Cogitri** [comment]
With the latest commit all tests pass for me.

File: src/item.rs
Comment (zeenix) [inline src/item.rs:38]: With 2.0.1 this shouldn't be needed as properties are cached lazily by default (as they were supposed to).

Comment (Cogitri) [reply src/item.rs:38]: Some tests still fail w/o disabling caching like https://github.com/hwchen/secret-service-rs/blob/b44a455626873283dfd52c9a099c69b4742d14ec/src/item.rs#L175. The `Item`'s label can still be read after deleting it and there's no call to `org.freedesktop.DBus.Properties.Get` before getting the label again.

File: Cargo.toml
Comment (bilelmoussaoui) [inline Cargo.toml:26]: you can use those directly from zbus iirc

Comment (Cogitri) [reply Cargo.toml:26]: AFAICS that isn't compatible with `zvariant_derive` (which isn't exported from `zbus`):

```
error[E0433]: failed to resolve: could not find `zvariant` in the list of imported crates
  --> src/proxy/collection.rs:58:41
   |
58 | #[derive(Debug, Serialize, Deserialize, Type)]
   |                                         ^^^^ could not find `zvariant` in the list of imported crates
   |
   = note: this error originates in the derive macro `Type` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Which happens even when I do `use zbus::zvariant`. I can remove `zbus_maros` though, thanks for the heads up.

File: src/item.rs
Comment (zeenix) [reply src/item.rs:38]: Interesting, this seems like a zbus bug then. I'll investigate soon but I good that you've a simple good workaround for now.

File: Cargo.toml
Comment (zeenix) [reply Cargo.toml]: Right I noticed this issue yesterday as well. IIRC, it works if you:

```rust
use zbus::zvariant;
use zvariant::Type;
```

Will see how we can solved this.

Comment (zeenix) [reply Cargo.toml]: Actually NM my last comment, that doesn't work either. :( However, for the workaround I'd suggest adding dep only only `zvariant` and then using macros from there directly:

```rust
use zvariant::Type;
```

macro crate only exists cause Rust wants macros in a separate crate.

Comment (Cogitri) [reply Cargo.toml]: Done, thanks

Comment (zeenix) [reply Cargo.toml]: Fix is [coming soon](https://gitlab.freedesktop.org/dbus/zbus/-/merge_requests/456/diffs?commit_id=5c1d4c9807ae56db236119a0bd7cfc2f91edf3bd) in the next zvariant release.

File: Cargo.toml
Comment (A6GibKm) [inline Cargo.toml:23]: Looking at the rest of the dependencies, this one should probably be "3.0.0", or "3.0" at worst. Ditto for `zbus`

**Review by complexspaces — APPROVED** [review]
All the tests now pass for me, even in multithreaded mode, so I think this is good to go. Thanks again for doing this.

**Comment by complexspaces** [comment]
The CI failure is unrelated. Its happening because the `zbus` crate family has bumped its unofficial MSRV to 1.54 with including the README in the Rustdoc documentation. Given that this crate is heavily reliant on `zbus` anyway, I'll bump CI's MSRV test to 1.54 in a followup to this.

File: Cargo.toml
Comment (complexspaces) [reply Cargo.toml]: This is correct to my knowledge as well. Specifying only a major version with no range exceptions will allow `cargo` to update as desired and pick the latest version with no dependency conflicts.

**Pull request merged**
